### PR TITLE
Feature/test accessmanager

### LINKED ISSUE #272: NullPointerException on header parse
I'm running into the following issue client side: https://github.com/square/okhttp/issues/2099, where I tell OkHttp to use `application/json;version=2`, but then the OkHttp client library converts this to `application/json;version=2; charset=utf-8`.

This ends up throwing the following exception server-side:

```
07-Nov-2022 19:35:38.693 WARNING [https-jsse-nio-8443-exec-3] cwms.radar.ApiServlet.lambda$init$10 error on request[3357625609381316046]: /cwms-data/timeseries
	java.lang.NullPointerException
		at cwms.radar.api.TimeSeriesController.deserializeTimeSeries(TimeSeriesController.java:400)
		at cwms.radar.api.TimeSeriesController.deserializeTimeSeries(TimeSeriesController.java:395)
		at cwms.radar.api.TimeSeriesController.create(TimeSeriesController.java:127)
		at io.javalin.apibuilder.CrudFunction$3.invoke$lambda-0(CrudHandler.kt:32)
		at io.javalin.apibuilder.CrudFunctionHandler.handle(CrudHandler.kt)
		at cwms.radar.security.CwmsAccessManager.manage(CwmsAccessManager.java:42)
		at io.javalin.http.JavalinServlet.addHandler$lambda-0(JavalinServlet.kt:96)
		at io.javalin.http.JavalinServlet$lifecycle$2$1$1.invoke(JavalinServlet.kt:43)
		at io.javalin.http.JavalinServlet$lifecycle$2$1$1.invoke(JavalinServlet.kt:43)
		at io.javalin.http.JavalinServletHandler.executeNextTask(JavalinServletHandler.kt:99)
		at io.javalin.http.JavalinServletHandler.queueNextTaskOrFinish$lambda-1(JavalinServletHandler.kt:85)
		at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:995)
		at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2137)
		at io.javalin.http.JavalinServletHandler.queueNextTaskOrFinish$javalin(JavalinServletHandler.kt:85)
		at io.javalin.http.JavalinServletHandler.executeNextTask$lambda-11$lambda-10(JavalinServletHandler.kt:119)
		at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:616)
		at java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:628)
		at java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:1996)
		at io.javalin.http.JavalinServletHandler.executeNextTask(JavalinServletHandler.kt:119)
		at io.javalin.http.JavalinServletHandler.queueNextTaskOrFinish$lambda-1(JavalinServletHandler.kt:85)
		at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:995)
		at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2137)
		at io.javalin.http.JavalinServletHandler.queueNextTaskOrFinish$javalin(JavalinServletHandler.kt:85)
		at io.javalin.http.JavalinServlet.service(JavalinServlet.kt:89)
		at javax.servlet.http.HttpServlet.service(HttpServlet.java:779)
		at cwms.radar.ApiServlet.service(ApiServlet.java:343)
		at javax.servlet.http.HttpServlet.service(HttpServlet.java:779)
		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227)
		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
		at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
		at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:197)
		at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
		at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:660)
		at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135)
		at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
		at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:687)
		at org.apache.catalina.authenticator.SingleSignOn.invoke(SingleSignOn.java:312)
		at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
		at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:360)
		at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:399)
		at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
		at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:893)
		at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1789)
		at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
		at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
		at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
		at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
		at java.lang.Thread.run(Thread.java:750)

```

Reading this post (though I'm far from knowledge on the subject to vet the answer) this format looks valid: https://stackoverflow.com/questions/29549299/how-to-split-header-values

### PR DESCRIPTION
**Pull request by MikeNeilson**
Initial code for setting up an AccessManager for the test environment.

### CODE REVIEW

**Review by rma-rripken — DISMISSED** [review]

**Comment by MikeNeilson** [comment]
Converted to draft, the integration tests now fail as I hadn't migrated the auth info in the tests.

File: cwms_radar_api/src/main/java/cwms/radar/datasource/DelegatingConnectionPreparer.java [hunk 1]
```diff
@@ -28,6 +38,7 @@ public DelegatingConnectionPreparer(ConnectionPreparer prep1, ConnectionPreparer
     public Connection prepare(Connection connection) throws SQLException {
         Connection retval = connection;
         for (ConnectionPreparer delegate : delegates) {
+            logger.info(delegate.getClass().getName());
```
Comment (rma-rripken) [inline cwms_radar_api/src/main/java/cwms/radar/datasource/DelegatingConnectionPreparer.java]: I think this info() statement should be something quieter.

Comment (MikeNeilson) [reply cwms_radar_api/src/main/java/cwms/radar/datasource/DelegatingConnectionPreparer.java]: Yep. Missed one. Thanks.

File: cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+            }
+        } catch(SQLException ex) {
+            logger.log(Level.WARNING,"Failed to retrieve roles for user",ex);
+        }
+        return roles;
+    }
+
+    private String checkKey(String key, Context ctx) {
+        try(Connection conn = dataSource.getConnection();
+            PreparedStatement setApiUser = conn.prepareStatement("begin cwms_env.set_session_user_direct(upper('q0hectest_pu')); end;");
```
Comment (rma-rripken) [inline cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java]: q0hectest_pu seems like it should be pulled out into a constant. Is that a new user for auth checks?

File: cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+            PreparedStatement checkForKey = conn.prepareStatement("select userid from cwms_20.at_api_keys where apikey = ?")) 
+        {
+            setApiUser.execute();
+            checkForKey.setString(1,key);
```
Comment (rma-rripken) [inline cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java:151]: I'll be curious to see all the db interactions that are happening for each request

File: cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java (hunk 2 above)
Comment (MikeNeilson) [reply cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java]: oy, thanks, that actually needs to be a "select user from dual" call.

File: cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java (hunk 3 above)
Comment (MikeNeilson) [reply cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java:151]: So far it's seemed pretty lightweight, if it starts to add up we can implement a form of session system were anyone can use what ever credentials have creds that don't need to round trip to the database for basic checks.

**Review by rma-rripken — DISMISSED** [review]

File: cwms_radar_api/build.gradle [hunk 4]
```diff
@@ -237,6 +252,7 @@ task integrationTests(type: Test) {
     jvmArgs += "-Djava.util.logging.config.file=$projectDir/logging.properties"
     jvmArgs += "-Dorg.apache.tomcat.util.digester.PROPERTY_SOURCE=org.apache.tomcat.util.digester.EnvironmentPropertySource"
     jvmArgs += "-Dradar.access.provider=KeyAccessManager"
+    jvmArgs += "-Doracle.jdbc.Trace=true"
```
Comment (rma-rripken) [inline cwms_radar_api/build.gradle]: Is this needed or left over?    I haven't played around with this -D, what does it do?  -Doracle.jdbc.Trace=true

**Review by rma-rripken — DISMISSED** [review]

**Comment by MikeNeilson** [comment]
That’s left over. It was supposed to give me more logging from the oracle drive as I was trying to track down an error. Need to remove it.


Also, this is still a draft PR, found additional database fixes are required.

Mike

From: rma-rripken ***@***.***>
Sent: Wednesday, January 25, 2023 12:44 PM
To: USACE/cwms-radar-api ***@***.***>
Cc: Mike Neilson ***@***.***>; Author ***@***.***>
Subject: [URL Verdict: Neutral][Non-DoD Source] Re: [USACE/cwms-radar-api] Feature/test accessmanager (PR #238)


@rma-rripken commented on this pull request.

________________________________

In cwms_radar_api/build.gradle<https://github.com/USACE/cwms-radar-api/pull/238#discussion_r1087146915>:

> @@ -237,6 +252,7 @@ task integrationTests(type: Test) {

     jvmArgs += "-Djava.util.logging.config.file=$projectDir/logging.properties"

     jvmArgs += "-Dorg.apache.tomcat.util.digester.PROPERTY_SOURCE=org.apache.tomcat.util.digester.EnvironmentPropertySource"

     jvmArgs += "-Dradar.access.provider=KeyAccessManager"

+    jvmArgs += "-Doracle.jdbc.Trace=true"

Is this needed or left over? I haven't played around with this -D, what does it do? -Doracle.jdbc.Trace=true

**Comment by MikeNeilson** [comment]
Okay, got the primary work done, if using the KeyAccessManager the latest unreleased (but is available under the latest-dev tag in the schema installer) is required due to the addition of some new functions and permission tweaks.

Next on the docket is a "combined access manager" and fixing up the rest of the tests. but this is too much change to wait for all that to be done.

File: cwms_radar_api/build.gradle (hunk 4 above)
Comment (MikeNeilson) [reply cwms_radar_api/build.gradle]: it was supposed to give more jdbc info to figure something out but it didn't work.

File: cwms_radar_api/src/main/java/cwms/radar/api/ClobController.java [hunk 5]
```diff
@@ -260,14 +262,17 @@ public Clob deserialize(String body, String formatHeader) throws JsonProcessingE
...
+        if (ContentType.equivalent(Formats.XMLV2,format)) {
             JacksonXmlModule module = new JacksonXmlModule();
             module.setDefaultUseWrapper(false);
             om = new XmlMapper(module);
-        } else if (Formats.JSONV2.equals(format)) {
+        } else if (ContentType.equivalent(Formats.JSONV2,format)) {
             om = JsonV2.buildObjectMapper();
         } else {
-            throw new FormattingException("Format is not currently supported for Levels");
+            FormattingException fe = new FormattingException("Format is not currently supported for Levels");
```
Comment (rma-rripken) [inline cwms_radar_api/src/main/java/cwms/radar/api/ClobController.java]: message should be for Clob not Levels

Comment (MikeNeilson) [reply cwms_radar_api/src/main/java/cwms/radar/api/ClobController.java]: Do'h, I had caught one other of those but missed that one.

Comment (MikeNeilson) [reply cwms_radar_api/src/main/java/cwms/radar/api/ClobController.java]: one of these days we should really move those to a properties file or something, like 90% of that text is duplicate in several controllers.

File: cwms_radar_api/src/main/java/cwms/radar/security/CwmsAccessManager.java [hunk 6]
```diff
@@ -153,14 +147,15 @@ private Set<RouteRole> getRoles(Principal principal) {
...
                 CwmsUserPrincipal cup = (CwmsUserPrincipal) principal;
                 roleNames = cup.getRoles();
-            } catch (ClassCastException e) {
+            //} 
+            /*catch (ClassCastException e) {
                 // The object is created by cwms_aaa with the cwms_aaa classloader.
                 // It's a CwmsUserPrincipal but it's not our CwmsUserPrincipal.
                 roleNames = callGetRolesReflectively(principal);
-            }
+            }*/
```
Comment (rma-rripken) [inline cwms_radar_api/src/main/java/cwms/radar/security/CwmsAccessManager.java]: Just remove the the try/catch if you're commenting out everything in the catch branch.   Or replace the reflexive stuff with a WARNING message that the cwms_aaa jar needs to be in the tomcat lib dir or whatever the specific fix is.

File: cwms_radar_api/src/main/java/cwms/radar/security/CwmsAccessManager.java [hunk 7]
```diff
@@ -169,7 +164,7 @@ private Set<RouteRole> getRoles(Principal principal) {
         }
         return retval;
     }
-
+/*
```
Comment (rma-rripken) [inline cwms_radar_api/src/main/java/cwms/radar/security/CwmsAccessManager.java]: just delete it if you're /* */ commenting it out.

Comment (MikeNeilson) [reply cwms_radar_api/src/main/java/cwms/radar/security/CwmsAccessManager.java]: yep, totally forgot about that section.

File: cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+            }
+        }
+
+        return retval;
+    }
+
+    private Set<RouteRole> getRoles(String user,String office) {
+        Set<RouteRole> roles = new HashSet<>();
+        try(Connection conn = dataSource.getConnection();
+            PreparedStatement setApiUser = conn.prepareStatement("begin cwms_env.set_session_user_direct(upper(?)); cwms_env.set_session_office_id(?);end;");
```
Comment (rma-rripken) [inline cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java]: you pulled the other statements into constants, why not all of them?

Comment (MikeNeilson) [reply cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java]: would you believe, laziness. But good to call that our. I will adjust.

File: cwms_radar_api/src/test/java/fixtures/TomcatServer.java [hunk 9]
```diff
@@ -121,6 +119,11 @@ public Realm getRealm(){
      */
     public void start() throws LifecycleException {
         tomcatInstance.start();
+        try {
```
Comment (rma-rripken) [inline cwms_radar_api/src/test/java/fixtures/TomcatServer.java]: why is the sleep needed?  Can you add a comment on what happens if the sleep isn't there or how to see if 5 seconds isn't long enough?

File: cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java (hunk 8 above)
Comment (MikeNeilson) [reply cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java]: good call, managed to leave a hardcoded office in one instance.

File: cwms_radar_api/src/test/java/fixtures/TomcatServer.java [hunk 10]
```diff
@@ -77,24 +76,23 @@ public TomcatServer(final String baseDir,
...
+            ExpandWar.delete(new File(existingRadar.getAbsolutePath()+".war"));
+            ExpandWar.copy(radar, new File(tomcatInstance.getHost().getAppBaseFile(),"cwms-data.war"));
         } catch( Exception err) {
             System.out.println(err.getLocalizedMessage());
         }
 
-        Context context = tomcatInstance.addWebapp(contextName, radar.toURI().toURL());
+        //Context context = tomcatInstance.addWebapp(contextName, radar.toURI().toURL());
 
         if(authValve != null && realm != null)
```
Comment (rma-rripken) [inline cwms_radar_api/src/test/java/fixtures/TomcatServer.java:91]: Can authValve just go away?  Seems like we don't use it anymore?

File: cwms_radar_api/src/test/java/fixtures/TomcatServer.java (hunk 9 above)
Comment (MikeNeilson) [reply cwms_radar_api/src/test/java/fixtures/TomcatServer.java]: actually I don't think it is. something was going on that was confusing me but if I remember right it was some other obvious thing like I had a file name slightly wrong so something wasn't loading.

File: cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+                String user = authorized(ctx, key, routeRoles);
+                prepareContextWithUser(ctx, user,key);
+            }
+            handler.handle(ctx);
+        }
+        catch(CwmsAuthException ex) {
+            logger.log(Level.WARNING,"Unauthorized login attempt",ex);
+            HashMap<String,String> msg = new HashMap<>();
+            msg.put("message",ex.getMessage());
+            RadarError re = new RadarError("Unauthorized",msg,true);
```
Comment (rma-rripken) [inline cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java:71]: There are minor formatting/spacing issues in this file

File: cwms_radar_api/src/main/java/cwms/radar/formatters/ContentType.java [hunk 12]
```diff
@@ -46,17 +47,30 @@ public String getCharset() {
...
+        /** We loop through instead of using contains key. 
+         *  Content-type parameter names are not case sensitive.
+         */
         for (Map.Entry<String, String> entry : parameters.entrySet()) {
             String key = entry.getKey();
-            if (key.equals("q")) continue; // we don't care about q for equals
-            if (!entry.getValue().equals(o.parameters.get(key))) return false;
+            if( "version".equalsIgnoreCase(key) ) {
+                return entry.getValue().equals(o.parameters.get(key));
+            }
```
Comment (rma-rripken) [inline cwms_radar_api/src/main/java/cwms/radar/formatters/ContentType.java:73]: I think this change deserves some tests in ContentTypeTest to demonstrate what we expect to be equals and what isn't.

File: cwms_radar_api/src/main/java/cwms/radar/formatters/ContentType.java [hunk 13]
```diff
@@ -88,4 +102,18 @@ public String toString() {
...
     }
+
+    /**
+     * Used for quick comparisons where we don't further need the content type.
+     * @param a
+     * @param b
+     * @return whether they are equivalent
+     */
+    final public static boolean equivalent(String a, String b) {
+        Objects.requireNonNull(a, "Cannot determine equivalency of null content-types");
```
Comment (rma-rripken) [inline cwms_radar_api/src/main/java/cwms/radar/formatters/ContentType.java:115]: Needs unit test.  tests should also demonstrate how this differs from equals

File: cwms_radar_api/src/test/java/fixtures/TomcatServer.java (hunk 10 above)
Comment (MikeNeilson) [reply cwms_radar_api/src/test/java/fixtures/TomcatServer.java:91]: yeah, but before I go that far I want to clean up other things.

File: cwms_radar_api/src/test/java/fixtures/RadarApiSetupCallback.java [hunk 14]
```diff
@@ -35,11 +39,11 @@ public class RadarApiSetupCallback implements BeforeAllCallback,AfterAllCallback
...
     private static CwmsDatabaseContainer cwmsDb;
-    private static TestRealm realm;
-    private static TestAuthValve authValve;
-    private static final String ORACLE_IMAGE = System.getProperty("RADAR.oracle.database.image", CwmsDatabaseContainer.ORACLE_18XE);
+
+    private static final String ORACLE_IMAGE = System.getProperty("RADAR.oracle.database.image", CwmsDatabaseContainer.ORACLE_19C);
     private static final String ORACLE_VOLUME = System.getProperty("RADAR.oracle.database.volume", "cwmsdb_radar_volume");
-    private static final String CWMS_DB_IMAGE = System.getProperty("RADAR.cwms.database.image", "registry.hecdev.net/cwms_schema_installer:21.1.1");
+    private static final String CWMS_DB_IMAGE = System.getProperty("RADAR.cwms.database.image", "registry.hecdev.net/cwms/schema_installer:latest-dev");
+
```
Comment (rma-rripken) [inline cwms_radar_api/src/test/java/fixtures/RadarApiSetupCallback.java:46]: Does this mean we're always running against the latest version of the installer or the latest version of the schema or both?  I think I'd rather depend on a specific version than latest.

Comment (MikeNeilson) [reply cwms_radar_api/src/test/java/fixtures/RadarApiSetupCallback.java:46]: That does mean the latest schema installer (basically what's in master). The tests currently do depend on it. But we should change it once a release that includes the changes is made.

File: cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java (hunk 8 above)
Comment (rma-rripken) [reply cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java]: Submit those fixes and this PR is prob good

File: cwms_radar_api/src/main/java/cwms/radar/formatters/ContentType.java (hunk 12 above)
Comment (MikeNeilson) [reply cwms_radar_api/src/main/java/cwms/radar/formatters/ContentType.java:73]: simple test added. I'll grant the names may not be the best but for the moment I was going for change the least amount of code at one time.

**Review by rma-rripken — APPROVED** [review]

**Pull request merged**